Added feature to check for unsaved files in a folder upon deletion.

kpatl1 opened the pull request:
**Summary**
I have added a function that checks if there are unsaved files within a folder. If so, it will open a modal asking if the user wants to actually delete the folder. On deletion it will then move to recycle bin. 
**Testing**
How did you test this change?
I followed the steps to recreate the feature request by joining a viya server, making a new folder with two unsaved files, and then try to delete the folder. After deleting it, I got the message if I truly did want to and then I had the option to continue or cancel. On continuation, the folder then moved to recycle bin. 

From:
https://github.com/sassoftware/vscode-sas-extension/issues/1136

scnwwu commented on client/src/connection/rest/SASContentAdapter.ts:
```diff
@@ -1,6 +1,6 @@
 // Copyright © 2024, SAS Institute Inc., Cary, NC, USA.  All Rights Reserved.
 // SPDX-License-Identifier: Apache-2.0
-import { FileType, Uri, authentication } from "vscode";
```
This file is an adapter to implement on REST service. This feature looks not necessarily specific to REST.
UI feature should be in UI component
https://github.com/sassoftware/vscode-sas-extension/blob/main/client/src/components/ContentNavigator/index.ts#L105
You can see warning message there. Please follow the pattern to make UI strings translatable.

scnwwu commented on client/src/connection/rest/SASContentAdapter.ts:
```diff
@@ -738,6 +742,36 @@ class SASContentAdapter implements ContentAdapter {
+    if (!isContainer(item)) {
+      return true;
+    }
+    const children = await this.getChildItems(item);
```
Looks like this can only get one level of children. VS Code native file explorer can track across multiple levels.
I'm concerned that it will take too long to traverse all children.

kpatl1 commented:
Added fixes. Put the logic into the UI and Service layer. Children are found recursively for the folder and added to a set. Open files are then referenced into the set to see if they are within it. If the file is dirty and in the children set, a popup will appear. Overall, time complexity boils down to O(N + M) I believe where N is the number of children and M being the number of files being open in the window.

scnwwu commented on client/src/components/ContentNavigator/ContentDataProvider.ts:
```diff
@@ -505,6 +505,46 @@ class ContentDataProvider
+
+    const descendants: ContentItem[] = [];
+    async function recurse(item: ContentItem) {
+      const kids = await this.model.getChildren(item);
```
Each getChildren call is an HTTP request. There can be many nested folders in real world so there can be many HTTP calls. Due to network latency, it can take several seconds to respond to user, which is not acceptable.

kpatl1 replied:
Would it be faster to do a BFS approach with a promise.all() so the HTTP requests are made concurrently?


```suggestion
public async checkFolderDirty(resource: ContentItem): Promise<boolean> {
  if (!resource.vscUri) {
    return false;
  }

  const descendants: ContentItem[] = [];
  const queue: ContentItem[] = [resource];


  while (queue.length > 0) {
    const currentBatch = queue.splice(0); // Dequeue all current level items
    const childrenArrays = await Promise.all(
      currentBatch.map(item => this.model.getChildren(item))
    );

    for (const children of childrenArrays) {
      for (const child of children) {
        descendants.push(child);
        if (getIsContainer(child)) {
          queue.push(child);
        }
      }
    }
  }


  const dirtyIds = new Set<string>();
  for (const doc of workspace.textDocuments) {
    if (!doc.isDirty) continue;
    const id = new URLSearchParams(doc.uri.query).get("id");
    if (id) dirtyIds.add(id);
  }

  for (const child of descendants) {
    if (!child.vscUri) continue;
    const childId = new URLSearchParams(child.vscUri.query).get("id");
    if (childId && dirtyIds.has(childId)) {
      return true;
    }
  }

  return false;
}

```

kpatl1 replied:
I agree. I am open to implementing this through the caching of the getTreeItem. However,  in the case that getTreeItem is not called, then the feature is inherently deployed with a "bug" built-in. But, as you said it is okay in some cases to not display the message we can implement this, and then continue to see if there is a way to improve it.

kpatl1 replied:
I have gone in and changed the logic to reduce the HTTP requests -- only using it as a fallback if the file is not already cached. Also added a depth check limit of 10 to prevent overflow. However, I am having trouble with it working with a file that is in a subfolder of a folder that is not saved when trying to delete the main folder. 

Ex. 
Test/
├── 1.sas
└── Test2/
    └── 2.sas

When trying to delete 2.sas, it does not show a warning.

scnwwu commented on client/src/components/ContentNavigator/index.ts:
```diff
@@ -523,5 +540,4 @@ class ContentNavigator implements SubscriptionProvider {
     }
   }
 }
-
```
this is unresolved. it's still showing an unnecessary removed line

scnwwu commented on client/src/components/ContentNavigator/index.ts:
```diff
@@ -107,30 +107,47 @@ class ContentNavigator implements SubscriptionProvider {
-                    ? Messages.FolderDeletionError
-                    : Messages.FileDeletionError,
-                );
+              if (resource.contextValue.includes("delete")) {
```
this is not resolved. it's not necessary to add this line in this change. this check was moved to line 106 above by a previous PR

scnwwu commented on client/src/components/ContentNavigator/ContentDataProvider.ts:
```diff
@@ -214,6 +217,11 @@ class ContentDataProvider
     const uri = await this.model.getUri(item, false);
 
+    // Cache the URI to parent mapping
+    if (item.parentFolderUri) {
+      this.uriToParentMap.set(item.uri, item.parentFolderUri);
+    }
```
Top level folder doesn't have a parentFolder.
Currently in `isDescendantOf`, as long as there's not a parentFolder, it will call `this.model.getResourceByUri`. And again the parentFolder is still empty. They're unnecessary network calls recursively until hit the 10 times limit. They should be avoided.
The fallback call is only used in case an item didn't go through this line. But if there's not a parentFolderUri here, it means this item really doesn't have a parent. Calling `this.model.getResourceByUri` is unnecessary and will result the same.
Perhaps something like

```suggestion
   const STOP_SIGN = '__STOP_SIGN__';
   this.uriToParentMap.set(item.uri, item.parentFolderUri ?? STOP_SIGN);
```

scnwwu commented on client/src/components/ContentNavigator/ContentDataProvider.ts:
```diff
@@ -505,6 +513,94 @@ class ContentDataProvider
+      if (currentParentUri === ancestorFolderUri) {
+        return true;
+      }
+
```
```suggestion
  if (currentParentUri === STOP_SIGN) {
    return false;
  }
```

scnwwu commented on client/src/components/ContentNavigator/ContentDataProvider.ts:
```diff
@@ -79,6 +80,8 @@ class ContentDataProvider
   public dropMimeTypes: string[];
   public dragMimeTypes: string[];
 
+  private uriToParentMap = new Map<string, string>();
```
I guess the cache should be cleared when [refresh](https://github.com/sassoftware/vscode-sas-extension/blob/main/client/src/components/ContentNavigator/ContentDataProvider.ts#L445)

kpatl1 reviewed (commented):
resolved comments, and tested functionality again.

Zhirong2022 commented:
It has no further response if trying to delete the folder which contains unsaved file after switch to a new Viya profile.
Steps to reproduce:
Suppose having two valid Viya connection
1.Sign in with one of the Viya server
2.Go to SAS Server, create a new file under a specified personal file folder and kept it unsaved
3.Switch to another Viya server
4.Delete the specified file folder which contains the unsaved file

scnwwu commented:
> > It has no further response if trying to delete the folder which contains unsaved file after switch to a new Viya profile. Steps to reproduce: Suppose having two valid Viya connection 1.Sign in with one of the Viya server 2.Go to SAS Server, create a new file under a specified personal file folder and kept it unsaved 3.Switch to another Viya server 4.Delete the specified file folder which contains the unsaved file
> 
> I believe that is because VSCode will no longer mark the file as dirty since you are reloading the connection. I am unsure is there is a way to work around this as it is a base issue coming from VSCode no?

Hi @kpatl1, @Zhirong2022 and I looked at it. I think it's not by VS Code. The dirty file is still dirty file. It's the underlying SAS server changed. It doesn't make sense to look up the parent of a file from a previous server on the new server. I would suggest to remove the fallback logic to fetch parent from server. If the cache doesn't contain the uri, it's acceptable to not pop up.

Zhirong2022 commented:
The raised issue has been fixed properly.
